sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala

```diff
@@ -0,0 +1,641 @@
+
+    RegexHexDigit(hexDigit)
+  }
+
+  private def isOctalDigit(ch: Char): Boolean = ch >= '0' && ch <= '7'
+
+  private def parseOctalDigit: RegexOctalChar = {
+    // \0n   The character with octal value 0n (0 <= n <= 7)
+    // \0nn  The character with octal value 0nn (0 <= n <= 7)
+    // \0mnn The character with octal value 0mnn (0 <= m <= 3, 0 <= n <= 7)
+
+    def parseOctalDigits(n: Integer): RegexOctalChar = {
+      val octal = pattern.substring(pos, pos + n)
+      pos += n
+      RegexOctalChar(octal)
+    }
+
+    if (!eof() && isOctalDigit(pattern.charAt(pos))) {
+      if (pos + 1 < pattern.length && isOctalDigit(pattern.charAt(pos + 1))) {
+        if (pos + 2 < pattern.length && isOctalDigit(pattern.charAt(pos + 2))
+            && pattern.charAt(pos) <= '3') {
+          parseOctalDigits(3)
+        } else {
+          parseOctalDigits(2)
+        }
+      } else {
+        parseOctalDigits(1)
+      }
+    } else {
+      throw new RegexUnsupportedException(
+        "Invalid octal digit", Some(pos))
+    }
+  }
+
+  /** Determine if we are at the end of the input */
+  private def eof(): Boolean = pos == pattern.length
+
+  /** Advance the index by one */
+  private def skip(): Unit = {
+    if (eof()) {
+      throw new RegexUnsupportedException("Unexpected EOF", Some(pos))
+    }
+    pos += 1
+  }
+
+  /** Get the next character and advance the index by one */
+  private def consume(): Char = {
+    if (eof()) {
+      throw new RegexUnsupportedException("Unexpected EOF", Some(pos))
+    } else {
+      pos += 1
+      pattern.charAt(pos - 1)
+    }
+  }
+
+  /** Consume the next character if it is the one we expect */
+  private def consumeExpected(expected: Char): Char = {
+    val consumed = consume()
+    if (consumed != expected) {
+      throw new RegexUnsupportedException(
+        s"Expected '$expected' but found '$consumed'", Some(pos-1))
+    }
+    consumed
+  }
+
+  /** Peek at the next character without consuming it */
+  private def peek(): Option[Char] = {
+    if (eof()) {
+      None
+    } else {
+      Some(pattern.charAt(pos))
+    }
+  }
+
+  private def consumeInt(): Option[Int] = {
+    val start = pos
+    while (!eof() && peek().exists(_.isDigit)) {
+      skip()
+    }
+    if (start == pos) {
+      None
+    } else {
+      Some(pattern.substring(start, pos).toInt)
+    }
+  }
+
+}
+
+/**
+ * Transpile Java/Spark regular expression to a format that cuDF supports, or throw an exception
+ * if this is not possible.
+ */
+class CudfRegexTranspiler {
+
+  val nothingToRepeat = "nothing to repeat"
+
+  def transpile(pattern: String): String = {
+    // parse the source regular expression
+    val regex = new RegexParser(pattern).parse()
+    // validate that the regex is supported by cuDF
+    validate(regex)
+    // write out to regex string, performing minor transformations
+    // such as adding additional escaping
+    regex.toRegexString
+  }
+
+  private def validate(regex: RegexAST): Unit = {
```

Comment:
nit: Do we want to possibly look at some rules that can re-write some of these? like if we see "()" as the regular expression can we replace it with ".*"?  I honestly don't know if that even would work, because I don't remember what java does in this case. This should probably be follow on work if we do want to look into this, because I don't want to hold this up from going in.

Reply:
I do think we should do this but I wanted to take this one step at a time and start off with simply falling back to CPU and then follow up with optimizations so that this PR doesn't become overwhelming to review. Ideally, I think we should follow up with one PR per specific optimization, so we can make sure that each one has comprehensive tests.

Reply:
That is fine with me. Being customer driven on what we pull in sounds good. After all most of these are corner cases, hey should be rare, and if we do add a modification step we need a lot of testing to really be sure it is doing the right thing.